File: src/java/magma/compile/rule/LazyRule.java
package magma.compile.rule;

import magma.compile.CompileError;
import magma.compile.Node;
import magma.compile.context.Context;
import magma.compile.context.NodeContext;
import magma.compile.context.StringContext;
import magma.option.None;
import magma.option.Option;
import magma.option.Some;
import magma.result.Err;
import magma.result.Result;

import java.util.function.Function;

public class LazyRule implements Rule {
    private Option<Rule> child = new None<>();

    @Override
    public Result<Node, CompileError> parse(String input) {
        return withChildSet(inner -> inner.parse(input), new StringContext(input));
    }

    @Override
    public Result<String, CompileError> generate(Node node) {
        return withChildSet(inner -> inner.generate(node), new NodeContext(node));
    }

    private <T> Result<T, CompileError> withChildSet(Function<Rule, Result<T, CompileError>> mapper, Context context) {
        return child.map(mapper).orElseGet(() -> new Err<>(new CompileError("Child not set", context)));
    }

    public void set(Rule child) {
        this.child = new Some<>(child);
    }
}

File: src/clang/magma/compile/rule/LazyRule.c
#include "LazyRule.h"
int __lambda0__(){return inner;
}
int __lambda1__(){return inner;
}
int __lambda2__(){return (CompileError("Child not set", context));
}
magma.result.Result<magma.compile.Node, magma.compile.CompileError> parse(String input){return withChildSet(__lambda0__.parse(input), StringContext(input));
}
magma.result.Result<String, magma.compile.CompileError> generate(magma.compile.Node node){return withChildSet(__lambda1__.generate(node), NodeContext(node));
}
magma.result.Result<magma.compile.rule.T, magma.compile.CompileError> withChildSet(magma.result.Result<magma.compile.rule.T, magma.compile.CompileError>(*mapper)(magma.compile.rule.Rule), magma.compile.context.Context context){return child.map(mapper).orElseGet(__lambda2__);
}
magma.compile.rule.void set(magma.compile.rule.Rule child){this.child = new Some<>(child);
}
